adr/0002-generate-extension-metadata-in-deployment-module.adoc

```diff
@@ -0,0 +1,123 @@
+
+Although it's not a serious problem, it would also be nice to allow people to use `CapabilityBuildItems` to register capabilities.
+At the moment, if authors do that, the information https://quarkus.io/guides/capabilities#capabilitybuilditem[will not be available to the dev tools], so we need to have lots of warnings in our documentation telling them not to.
+
+== Decision Drivers
+
+* Anything we do should minimise work for extension authors
+* Anything we do should minimise breaking changes, especially since many quarkiverse extensions will be using back-level versions of the Quarkus extension plugin
+* We should think carefully about the right information layout, so that we get the right information in the right long-term place, and only need to do this disruptive re-arrangement once
+
+== Considered options
+
+- *Manual creation of metadata*. This is tedious for users and error-prone; we could not assume any manually-managed metadata was complete. It's also annoyingly non-DRY; we'd be asking extension owners to write the same information down in two places (their code, and a metadata file). It does have the advantage that it's non-breaking.
+- *Running builds twice*. Because the circular dependency is a soft one, it works perfectly well to run a build to generate the deployment module's files, and then run a second build to pull those into the runtime module's metadata generation. Doing this would be a simple, non-breaking, change to release scripts. However, it is error-prone (if scripts are incorrectly run once-only, we just miss metadata), lengthens build times, breaks the Develocity by introducing hidden cross-module dependencies, and is generally Just Wrong.
+- *Doing nothing*. This is cheap, and non-breaking. But it means we miss an opportunity to take advantage of the fact that, at build time, we know a lot about what extensions can do and would like to make that information available to Quarkus users.
+- *Move (most) metadata generation to the deployment module* We can eliminate the circular dependency by not generating information which isn't particularly related to the runtime module in the runtime module. If the deployment module 'hosts' the `quarkus-extension.yaml`, it can harvest information from the extension's build steps and build items. This is architecturally clean, unlocks a range of capabilities in the future, and is highly disruptive in the present.
+
+== Proposal
+
+Moving metadata generation to the 'right' place is the architecturally correct thing to do.
+
+image::images/0002-resolution.svg[the proposed resolution]
+Source: https://excalidraw.com/#json=JMERR95gRufcmnUpWy6A_,Qd9efiR40LR3wv82CxyLKg
+
+So how do we minimise the impact?
+What needs to happen for it to work?
+
+* The registry reads `quarkus-extension.yaml` from the deployment module, not the runtime module. An alternate implementation would be to publish a separate `-metadata` artifact and the tooling could read that artifact.
```

Comment:
any reason registry is called out here ?

should it not be *any* tooling (registry, our plugins, IDE etc.) that should read from the deployment module?
meaning it is no longer enough to look in the classpath but will need to lookup the matching "-deployment" ?

Reply:
Actually, I realised it should say 'registry upload' (https://github.com/quarkusio/jbang-catalog/blob/main/catalog_publish.java#L318). By the time it hits the registry it's pure metadata, with no enclosing jar. 

I think it’s useful to call it out the registry, since it’s the main thing which consumes and dispenses metadata. It's also the tooling we control so can directly influence. Other tooling, like IDEs, might choose to take its chances and only support the new model, or take its chances and accept the ‘old’ data.

But I'll update with the correction about the upload, and also include other tools.